Fixing type errors for nn.conv

vivekmig opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41939 Fixing type errors for nn.conv**

Differential Revision: [D22729823](https://our.internmc.facebook.com/intern/diff/D22729823)

vivekmig committed:
Fixing type errors for nn.conv

[ghstack-poisoned]

vivekmig committed:
Update on "Fixing type errors for nn.conv"




[ghstack-poisoned]

vivekmig commented on torch/nn/modules/conv.py:
```diff
@@ -579,8 +579,8 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
-    def _output_padding(self, input, output_size, stride, padding, kernel_size):
-        # type: (Tensor, Optional[List[int]], List[int], List[int], List[int]) -> List[int]
+    def _output_padding(self, input: Tensor, output_size: Optional[List[int]], stride: List[int],
+                        padding: List[int], kernel_size: List[int]) -> List[int]:
```
Kernel_size, stride, and padding passed here are actually Tuple[int, ...], but Torchscript doesn't seem to support arbitrary length tuples. Based on #14775, seems like it's necessary to type these as List[int] for torchscript to convert arguments to lists.
This requires changing the attribute type to Any above to avoid type mismatch errors.

malfet replied:
One can use `cast` to convert `List` to `Tuple`

malfet commented on mypy.ini:
```diff
@@ -5,6 +5,7 @@ warn_unused_configs = True
 warn_redundant_casts = True
 show_error_codes = True
 check_untyped_defs = True
+allow_redefinition = True
```
It sounds wrong to allow it globally, perhaps it can be limited to just one file or even better, to just one line, by using `#type: ignore[no-redef]` , as for example done here:
https://github.com/pytorch/pytorch/blob/1b55e2b043b0be0f99de20477f4135ca4e94c3d0/torch/futures/__init__.py#L16

vivekmig replied:
The issue is due to cases like this `kernel_size = _single(kernel_size)`, where int or tuple of int is converted to a tuple, which occurs at least a few times in this file. I think this pattern could be common, but I can limit it for this file for now.

malfet commented on torch/nn/modules/conv.py:
```diff
@@ -23,9 +23,9 @@ class _ConvNd(Module):
-    kernel_size: Tuple[int, ...]
-    stride: Tuple[int, ...]
-    padding: Tuple[int, ...]
+    kernel_size: Any
+    stride: Any
+    padding: Any
```
Can you please explain why? `Tuple` or `Sequence` sounds like a right type here

vivekmig replied:
Tuple is the 'correct' type here, the issue with setting this to Tuple is it mismatches with _output_padding, which is typed with a List[int]. We can set this to Tuple[int, ...] and cast every call to _output_padding with List[int], but this unfortunately fails when scripting. Typing with Any avoids conflict with the signature of _output_padding

malfet commented on torch/nn/modules/conv.py:
```diff
@@ -37,14 +37,14 @@ class _ConvNd(Module):
-                 stride: _size_1_t,
-                 padding: _size_1_t,
-                 dilation: _size_1_t,
+                 kernel_size: Tuple[int, ...],
+                 stride: Tuple[int, ...],
+                 padding: Tuple[int, ...],
+                 dilation: Tuple[int, ...],
```
Again, why this change is needed? `padding=2` seems like a valid syntax, isn't it?

vivekmig replied:
Yes, child classes can take ints, but they are always converted to tuple prior to initialization of the parent construction, e.g. Conv1d calls `kernel_size = _single(kernel_size)`

vivekmig committed:
Update on "Fixing type errors for nn.conv"


Differential Revision: [D22729823](https://our.internmc.facebook.com/intern/diff/D22729823)

[ghstack-poisoned]
